### DISCUSSION TITLE
Can I have default providers on LoggingEventCompositeJsonEncoder?

### DISCUSSION THREAD

**Discussion by zmes50416** [logfellow/logstash-logback-encoder#857] [Q&A] [answered]
According to the [document](https://github.com/logfellow/logstash-logback-encoder/tree/logstash-logback-encoder-7.2#composite-encoderlayout) written:
> No providers are configured by default in the composite encoders/layouts. You must add the ones you want.

I am using the default LogstashEncoder, and if I want to modify the message column with some mdc value.
Does this means my only option is to use a LoggingEventCompositeJsonEncoder and add all providers by myself?

My logbook configuration will need to change from a simple three line xml
```xml
<appender name="simple" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LogstashEncoder" />
</appender>
```
To something like this? 
```xml
<appender name="customized_one" class="ch.qos.logback.core.ConsoleAppender">
        <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
            <providers>
                <threadName />
                <timestamp />
                <version />
                <loggerName />
                <logLevel />
                <logLevelValue />
                <mdc />
                <stackTrace />
                <pattern>
                    <pattern>
                        {
                        "message": "%mdc{custom_value} %message"
                        }
                    </pattern>
                </pattern>
            </providers>
        </encoder>
</appender>
```

I don't even know if I miss some origin fields or not... Or there is another way to do it?

**Comment by brenuart** [comment]
~~`LogstashEncoder` is kind of a "read-only" set of predefined encoders. Although you can customise some of them, you cannot add extra ones to the predefined list. As you noticed, your only option so far is to start with an empty `LoggingEventCompositeJsonEncoder` and add the encoders you need.~~ **(this is not correct - see further in this thread for more information)**

Using a LogstashEncoder is similar to the following configuration:
```xml
  <encoder class="net.logstash.logback.encoder.LoggingEventCompositeJsonEncoder">
        <providers>
                <timestamp />
                <version />
                <message />
                <loggerName />
                <threadName />
                <logLevel />
                <logLevelValue />
                <callerData />
                <stackTrace />
                <mdc />
                <tags />
                <logstashMarkers />
                <arguments />
        </providers>
  </encoder>
```

I recon the documentation is not very clear on the subject and we will do our best to take some time to review it in the near future. Sorry about that.
Feel free to come back with more questions if you need more help.

**Reply by zmes50416** [reply]
Thank you for the feedback and providing answer with detail.

**Reply by brenuart** [reply]
Hi @zmes50416 - sorry but I was wrong in my previous answer :( There is indeed a way to achieve what you are looking for. Please see  my next answer for more information.

**Comment by brenuart** [comment, marked as answer]
To add extra JsonProviders to those already registered by LogstashEncoder, you have to use the `<provider>` keyword as shown below:

```xml
<encoder class="net.logstash.logback.encoder.LogstashEncoder">
    <!-- Add a new provider after those than come with the LogstashEncoder -->
    <provider class="net.logstash.logback.composite.loggingevent.LoggingEventPatternJsonProvider">
        <pattern>
          {
             "message": "%mdc{custom_value} %message"
          }
        </pattern>
    </provider>

    <!-- Disable the default message provider -->
    <fieldNames>
        <message>[ignore]</message>
    </fieldNames>
</encoder>
```

You can add several additional JsonProviders using multiple `<provider>` entries. They will appear just after the default providers registered by the LogstashEncoder.

In this example, the pattern provider produces a "message" JSON field that will conflict with the message field already produced by the MessageJsonProvider registered by the LogstashEncoder. Different options to avoid the conflict:
- you configure the LogstashEncoder to use a different field name (using the [fieldNames](https://github.com/logfellow/logstash-logback-encoder#customizing-standard-field-names) configuration property)
- you disable the message provider that comes with the encoder (that's the option illustrated in the example above)
- you use a different field name in your pattern

Note: using the `<providers>` configuration property is not allowed and will be silently ignored (an error status should be printed but it will unfortunately not be visible because of #859).  
I recon the documentation is not very clear on the subject and the API is a bit cumbersome and hard to use... Sorry about that. We are working on it...

**Reply by zmes50416** [reply]
Thank you for your time to do further investigation. This one looks much cleaner to me 👍🏼
